crates/red_knot/src/lint.rs

```diff
@@ -191,6 +192,34 @@ fn lint_bad_overrides(context: &SemanticLintContext) -> QueryResult<()> {
     Ok(())
 }
 
+fn lint_unspecified_encoding(context: &SemanticLintContext) -> QueryResult<()> {
+    // symbol we want to match against
+    let target = resolve_global_symbol(
+        context.db.upcast(),
+        ModuleName::new("tempfile"),
+        "TemporaryFile",
+    )?;
+
+    for (symbol_id, definition) in context.symbols().all_definitions() {
```

Comment:
I'm not sure that iterating definitions is actually right for this lint rule. I don't think we really care if it's a definition or not, we'd want to detect all calls to e.g. `tempfile.TemporaryFile` that don't specify an encoding. So I think this lint rule is a better fit for an AST traversal (which we don't currently have an example of in the red-knot lint rules).

I think the algorithm should be, walk all calls, infer type of the `.func` of the call, check if its `tempfile.TemporaryFile`, if so, check the arguments and see if an encoding is specified or not.

Reply:
Yeah, thought this seemed awkward-- iterating over definitions is great for the type inference algorithm but totally nonsense for this rule. Should I go ahead and use `ruff_python_ast::visitor` since the AST type used here in redknot is the same as that of ruff?

Reply:
Yes, you can use `ruff_python_ast::visitor`. Longer term we will want to do something more efficient, e.g. where we do a single AST traversal that lint rules can register callbacks on. But for now given this is our first prototype lint rule doing an AST traversal, it's fine IMO to just traverse internal to the rule.